arcflow/utils/bulk_import.py

```diff
@@ -128,8 +172,122 @@ def csv_bulk_import(csv_directory=None, load_type='ao', only_validate='false'):
+
+        bulk_import_report.append(file_import_report)
         print(json.dumps(import_job, indent=4))
+    return bulk_import_report
+
+def save_report(path, report_list, validate_only):
+    """Function to create and save reports for tracking bulk imports."""
+    current_datetime = datetime.now().strftime('%Y-%m-%d-%H%M%S')
+    action = "validate" if validate_only else "import"
+    suffix = current_datetime + "_" + action
+    report_file_name_stem = Path(path).name + "_report_" + suffix
+    report_text_file_name = report_file_name_stem + ".txt"
+    report_save_path = os.path.join(path, "reports")
+
+    if not os.path.exists(report_save_path):
+        os.makedirs(report_save_path)
+
+    txt_report_save_path = os.path.join(report_save_path, report_text_file_name)
+    with open(txt_report_save_path, 'w', encoding='utf-8') as report:
+        print("Import Job Info", file=report)
+        json.dump(report_list, report, indent=4)
+
+    report_csv_file_name = report_file_name_stem + ".csv"
+
+    fieldnames = ['identifier','ead_id','repo_id', 'rid','only_validate','type','resource_id','error','results_status','results_warnings','results_id','results_uri',"csv_issue_count"]
+    
+    csv_report_save_path = os.path.join(report_save_path, report_csv_file_name)
+    with open(csv_report_save_path, "w", newline="", encoding='utf-8') as csvfile:
+        writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
+        writer.writeheader()
+        for row in report_list:
+            writer.writerow(row)
+
+def check_job_status(client, repo_id, job_id):
+    """Function to check whether a job has completed (and thus output files are ready)."""
+    while True:
+        job_status_response = client.get(f'/repositories/{repo_id}/jobs/{job_id}')
+        job_status = job_status_response.json()['status']
+
+        if job_status == 'completed':
+            print(f"Job {job_id} completed successfully!")
+            return True
+        elif job_status == 'failed':
+            print(f"Job {job_id} failed.")
+            return False
+        else:
+            pause = 15
+            print(f"Job {job_id} is still {job_status}. Waiting {pause} seconds...")
+            time.sleep(pause)
+
+def retrieve_job_output(path, report_list):
+    """Function to retrieve and save last created output files for each job in the bulk import."""
+    client = __get_asnake_client()
+    for row in report_list:
+        if "results_id" not in row:
+            continue
+        repo_id = row["repo_id"]
+        job_id = row["results_id"]
+        identifier = row["identifier"]
+        if not check_job_status(client, repo_id, job_id):
+            print(f"Error downloading files for job {job_id}")
+            continue
+        job_files = client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files").json()
+        output_file_id = max(job_files)
+        response = client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files/{output_file_id}")
```

Comment:
Since `retrieve_job_output` modifies the output of `csv_bulk_import`, I recommend to put `retrieve_job_output` inside `csv_bulk_import` to maintain logical cohesion, and add an `asnake_client` parameter to `retrieve_job_output` (this also eliminates the need to instantiate an additional `ASnakeClient`).

Part 4 of 4
```suggestion
def retrieve_job_output(path, report_list, asnake_client):
    """Function to retrieve and save last created output files for each job in the bulk import."""
    for row in report_list:
        if "results_id" not in row:
            continue
        repo_id = row["repo_id"]
        job_id = row["results_id"]
        identifier = row["identifier"]
        if not check_job_status(client, repo_id, job_id):
            print(f"Error downloading files for job {job_id}")
            continue
        job_files = asnake_client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files").json()
        output_file_id = max(job_files)
        response = asnake_client.get(f"/repositories/{repo_id}/jobs/{job_id}/output_files/{output_file_id}")
```